stan/math/opencl/kernel_cl.hpp

```diff
@@ -125,62 +133,48 @@ struct global_range_kernel {
     auto f = make_functor();
-    cl::EnqueueArgs eargs(opencl_context.queue(), global_thread_size);
-    f(eargs, args...).wait();
+    std::vector<cl::Event> kernel_events
+        = vec_concat(select_events<Args>(args)...);
+    cl::EnqueueArgs eargs(opencl_context.queue(), kernel_events,
+                          global_thread_size);
+    cl::Event kern_event = f(eargs, get_kernel_arg(args)...);
+    // This is just a dummy to set off the variadic function
+    int dummy[] = {0, (assign_event<Args>(args, kern_event), 0)...};
+    return kern_event;
   }
-};
-/**
- * Creates functor for kernels that need to define both
- *  local and global work size.
- * @tparam Args Parameter pack of all kernel argument types.
- */
-template <typename... Args>
-struct local_range_kernel {
-  const kernel_functor<Args...> make_functor;
-  /**
-   * Creates kernels that need access to defining the global thread
-   * size and the thread block size.
-   * @param name The name for the kernel
-   * @param source A string literal containing the code for the kernel.
-   * @param options The values of macros to be passed at compile time.
-   */
-  local_range_kernel(const char* name, const char* source,
-                     const std::map<const char*, int>& options = {})
-      : make_functor(name, {source}, options) {}
-  /**
-   * Creates kernels that need access to defining the global thread
-   * size and the thread block size.
-   * @param name The name for the kernel
-   * @param sources A std::vector of strings containing the code for the kernel.
-   * @param options The values of macros to be passed at compile time.
-   */
-  local_range_kernel(const char* name, const std::vector<const char*>& sources,
-                     const std::map<const char*, int>& options = {})
-      : make_functor(name, sources, options) {}
+
   /**
    * Executes a kernel
    * @param global_thread_size The global work size.
    * @param thread_block_size The thread block size.
    * @param args The arguments to pass to the kernel.
    * @tparam Args Parameter pack of all kernel argument types.
    */
-  auto operator()(cl::NDRange global_thread_size, cl::NDRange thread_block_size,
-                  Args... args) const {
+  auto operator()(
+      cl::NDRange global_thread_size, cl::NDRange thread_block_size,
+      const typename internal::to_matrix<Args>::type&... args) const {
     auto f = make_functor();
-    cl::EnqueueArgs eargs(opencl_context.queue(), global_thread_size,
-                          thread_block_size);
-    f(eargs, args...).wait();
+    std::vector<cl::Event> kernel_events
+        = vec_concat(select_events<Args>(args)...);
+    cl::EnqueueArgs eargs(opencl_context.queue(), kernel_events,
+                          global_thread_size, thread_block_size);
+    cl::Event kern_event = f(eargs, get_kernel_arg(args)...);
+    // This is just a dummy to set off the variadic function
+    int dummy[] = {0, (assign_event<Args>(args, kern_event), 0)...};
```

Comment:
I tried so hard to find a different way to do this and couldn't :'(

Reply:
samezess :(

It's not the worst, it's just that it's unintuitive